### DISCUSSION TITLE
a new user / trying to cut return of an intersection on 2 halves for a process

### DISCUSSION THREAD

**Discussion by gilboonet** [zalo/CascadeStudio#86] [Q&A] [answered]
Hello, I'm a new user from Belgium. Long time user of OpenJSCAD that I use to help me build cardboard structures, I was recently amazed by Cascade Studio and started to use it. I already built an object from it and it helped me design another one that I finished by hand with Wings 3d. Now I'm trying to make a very important thing for my crafting process : when I build a cardboard furniture, I start by creating its skeleton that is made of two pairs of cross pieces slices. Here's what I have : 
![Capture d’écran de 2021-03-16 15-04-15](https://user-images.githubusercontent.com/22082873/111324618-3a241d00-866b-11eb-97ab-16099a2f94e7.png) The volume is a simplified armchair (on the top left). The skeleton is made of vertical and horizontal slices (top right and bottom left). On the bottom right are the intersections of all slices. What I want to do is cut each of those intersections in two (a top one and a bottom one), then subtract the bottoms to the vertical slices and the tops to the horizontal slices, so that the skeleton can be easily assemble by combining those slices. But I don't see how I can get the needed data to do that, something like a bounding box. Here the intersections look simple, but some volumes can have a more complex set of intersections. I took a look at the source, mostly StandardLibrary.js, and also at all the scripts I could find, but didn't see what I can do. I'm not a professional js nor 3d developer, and I'm really grateful to you to make such a nice tool available.  
https://zalo.github.io/CascadeStudio/?code=jZPBbtswDIbvfgoeJVRJ5NhIM2S9FGixAcul6XoxfDAcuXJmyIOtZNmKvftIWUltzwN2iJ2Qv35%2BJJXiaHJb1gaeVG4z81opVgnQAnJlrGrgDoqsahV%2FCwAqZWGHEaN%2BwO6bsrlmiRQgUz7HLMCX0qjnmiXVdEyPY3I6dj37YPaMzx%2BzXDG%2BCVyoLFgHxgnkuclMW2VWsWRWLZYCZhqfMhWwwwPQKHtsDOw2we%2BguPRpG9%2FgWcBPAb84UGte2jP0afQaT6YbCCfXgEZyn7UKYR7OtjnuFUO3rwYrscQRYz1YC0A4wN5m%2BJZc9DLxusvQp5eJlmIZY9RnJQwOObv4337x1RAzqcu7yQqIVmlAwyTwx9rsh1NcU1Ga33199ggxdUrqLUqfaut0oZBO9gEdfbM0BeEsU34pYHWIh5J0A4sF4mUm16qFTz7n59ZfYijRV9DLMywxEAp8kmdR4%2B7waAl1AcksWlH%2FMcrdl5S%2FUcH592Or2We6I61yK2fJVvSrSFE6%2BI6AEy74VdrTFPCLz%2F0N3LH2gR0s7meKWIpbgayIHEUCVxTfOubTfzCXfuAdxIC51Kd35r4HThow%2F9KjuAO5wddHN6hKmVer8ffNjft7k%2BzQSQ4kOb1LDl4CVG2KFv2S0vGFySF1NwAQEBHHxYfGrvbgCqy6eyq9WZm6Tidcxh0M9wKri4m%2BmozLXCRbuq9%2FAA%3D%3D&gui=q1ZKzs8tyM9LzSvxS8xNVbJSSk4sTk5MSQ3LTC1X0lHyTS3OCEotVrIy0DOE84IS89KBSqMN9AwMdYxidZScE5MzUu2VrEqKSlN1lIISUzJLgVqMDWBsmAYjAx0Tg9haAA%3D%3D

**Comment by gilboonet** [comment]
Here's an example of the output I have with OpenJSCAD
![Capture d’écran de 2021-03-18 09-05-35](https://user-images.githubusercontent.com/22082873/111593113-e67c1580-87c9-11eb-9cd5-865cded27f3a.png)

**Comment by gilboonet** [comment]
I tried to write a function to get bounding box of a shape, so I could measure the intersections and split them in two halves... I had some troubles to find my way looking into the code, but I think i have something that works...
[see here](https://zalo.github.io/CascadeStudio/?code=fZJBj5swEIXv%2FIq3NyMh0lbqJWn2sG1X7a3abNttEaocMmyQHBuBqYBs%2FnvHxlRIqYoAe56%2FmXmMWK3wnVRhTgRr8F62hTwQdrY7VOYGwCdqCJKf1iFdS2WnUHa6sJXR7TparYDHRupWSUsiTvBgbNjtCqn85qtm1m0%2BVGXJ9XTh5c%2FaUtOSryTiSJGFxBa7%2BsiMePM23nhtz9oz2TvT6UMrJKsFdzaKUmWeRbbPXuUJ9tnrPOejaLa2SGmPsqb4HAF8%2B4qnSnPRc7%2BG7pRKMKyndZyES8Kgu%2FYn2f8X5IbAvWk%2ByuJ4L%2FmzfK8EotIH6hOUrMXY3uLsKwbyGzWWeuEOGf3towU1may15damSO8eqP71aIxK68ZYY4ea0i%2FaznmbKQWLVOeZ09MnN%2BUhBD9cMIbgp%2BBZLbKqEuLGjSXt8fIC0eMdpjCOg7xFv7miB08PMz0E2vUcrunR0%2BNMj4F2pkY%2FyQUu%2B79WbjGFDvfytRWWZyuBHgL9Lyssz1YCPQY6WHHsJY7caxM1ZLtGI8umQSThU8M68s%2BXTb6S0DmsY55vossf&gui=q1ZKzs8tyM9LzSvxS8xNVbJSSk4sTk5MSQ3LTC1X0lHyTS3OCEotVrIy0DOE84IS89KBSqMN9AwMdYxidZScE5MzUu2VrEqKSlN1lIISUzJLgVqMDWBsmAYjAx0Tg9haAA%3D%3D)
```

let a = Sphere(25);
let b = getBounds(a);
console.log([b[0], b[1]]);

function getBounds(shape){
  
  let bmin = {x: null, y:null, z: null},
      bmax = {x: null, y:null, z: null};

  ForEachFace(shape, (index, face) => {
    ForEachVertex(face, (vertex) => {
      let pnt = oc.BRep_Tool.prototype.Pnt(vertex);       
      let x = pnt.X(), y = pnt.Y(), z = pnt.Z();
    
      if (!bmin.x || (x < bmin.x))bmin.x = x;
      if (!bmin.y || (y < bmin.y))bmin.y = y;
      if (!bmin.z || (z < bmin.z))bmin.z = z;

      if (!bmax.x || (x > bmax.x))bmax.x = x;
      if (!bmax.y || (y > bmax.y))bmax.y = y;
      if (!bmax.z || (z > bmax.z))bmax.z = z;
    })
  });
return [[bmin.x, bmin.y, bmin.z], [bmax.x, bmax.y, bmax.z]];
}
```

**Comment by gilboonet** [comment]
With the getBounds function() (that needed to have !bmin.x replace by (bmin.x === null) ), it didn't take me long to be able to have my two halves. Now I only need to render 2d slices and it will be a functional script. That's already impressive to be able to do that so easily.
see [here](https://zalo.github.io/CascadeStudio/?code=jVVNb%2BM2EL3rV8yRxNKObKtBdr3eg9EEW6ABisQbtBWEQpbpiKlKGTLtlZXNf%2B8MSdnyR4oebEoz7828GZKj5UZnRpUaHmRmUv1cSFYIyAVkUhtZwQSWabGW%2FDUAKKSBR7Ro%2BR0e%2F5Ymy1kcCggT3kcvwK9Ky1nJ4uKyLT%2B1hZdte%2B6tXjDev0szyfg4sCa1ZE4YJyGzKtXrIjWSxb3iaiigl%2BN%2FmAh4RAJU0mwqDY%2Fj4C1YtnWayhdYC9gJaDhQaR7aCejdGOu0M64hnKIewj5LMy03erFm6zxduX75ls3%2FURrFvtafQG8KjLH75NbGGd6ELQ0QmNb%2FCbRNuCur2zTLbVtsLgFM6YWsSVkmOUy%2B2JL2yCdZGVkzciJ0a986KCdypQ2mLrP%2B9EGu%2FpqVZdFfVaUpzW4l%2B79p0%2FLGjgIdKmlGev93xlGyf%2FmDXhr%2F8iftX4ellsAYtaWP3MnEFsfhxw9gNXwG5%2BC8BUA9PuPtTni7lrfzPNKxO%2Bc1J7ym5TWeR5Ibf9j2xLS%2BIPQLOAcRHeBMKJrPhXrezvMuCUXzuVDPazzPCyXWGw%2Fobxz4UxzHrnXCt8SvdJZjp1V4DX5tksSeZtrNabqWGPq2NtVmIRlG%2FqbxgLPYpsLbAzcC8KoB3tQeriEXHU904zz063hGQzGM0Oq9IRyRbLjo%2FXjRPiB6Euu3c0LA6DoJaDSQ8LtSL45nwg0lpWkwLWsvIaJ7S%2Bh7hD6UxuIGIrSwjxjRF0tdEDZkwtsEJh8gKU7GcHWF8lKd5XINX73P9607kgYhxhW0eA1DNAwE%2FtuYnZFkY31lir%2B2k%2BgXGnJraQEsvhfdwChT0ZgUPi1P8FrSMCpxtqEYBeUS4t4ooo5GKOAaH6IEg2MJ%2FdVmnbNOSse1RWwvFfjkfecFutq6BdricD%2FfqfDp%2F1ao%2FI64rJfr%2B0lgXVjeaCSikS1ue1zc0764vQ5FOZnJKfJ7Spw3ObRF5dtDW7pY3HxAv2uQmlkQbs3UoTtyJxCOcfls%2B19I%2FWxyfP%2FwwX4lCPbiIC8E2R4gLx7i5qyiQG0Fg1jZ%2Fgzil8QPV%2FutQcjhU6RUx6UUOY9rVXiOLKDT%2B3kcJjE13z4M2odh4vd4jjY09Dq4AeF6e7yDDL1lmPCrIefJkZQcpfyslktZSY3fMdKBYuYtSM3cTiIQ98GZpq1pbkFvwdGJcC0%2Bbp%2FtsGNTo1RynNTZaMN81u7dvXYDJvRNVnQe3sl4uqc%2BY34ho982NTvPiMmu24T5IeGpphZzT%2Fcr%2BBc%3D&gui=q1ZKzs8tyM9LzSvxS8xNVbJSSk4sTk5MSQ3LTC1X0lHyTS3OCEotVrIygrODEvPSgQqjDfQMDHWMYnWUnBOTM1LtlaxKikpTdZSCElMyS4EajA1gbJgGIwMdE4PYWgA%3D)   
![Capture d’écran de 2021-03-20 22-32-46](https://user-images.githubusercontent.com/22082873/111886395-7d64ef80-89cd-11eb-8228-2e62fecff07f.png)

**Comment by gilboonet** [comment, marked as answer]
Well, I'm almost done. I now have the slices aligned as I need for my process (I mean cut them), but there's no option to save to a 2d format (svg or dxf), so I saved to .obj. First I tried to save the model as wire, in order to save it to a 2d format, but I finally saved to .obj and got a hollow model. Then I save the slices as 3d, but It's not a handy format for me. Is there a way to have my slices saved to a 2d format ?
 
![Capture d’écran de 2021-03-21 18-34-45](https://user-images.githubusercontent.com/22082873/111915664-bd80ac80-8a77-11eb-9fed-f6046184992b.png)
![Capture d’écran de 2021-03-21 18-30-01](https://user-images.githubusercontent.com/22082873/111915666-beb1d980-8a77-11eb-8809-f38f5b228008.png)

**Reply by gilboonet** [reply]
I mark this as answered as for the moment Cascade Studio doesn't save to any 2d format.

**Reply by zalo** [reply]
Apologies; OpenCascade itself doesn't save to any 2D formats, so there's nothing for me to expose...

If you're interested in simple SVG export, you may be able to traverse the edges/vertices of your shapes at the end, and print them into the SVG format.

**Comment by udif** [comment]
Wow, you really went into the rabbit's hole with this one...
I've tried looking at the OC documentation looking for bounding box functions, but got nowhere.
Thanks for sharing this!
